server/src/main/java/org/opensearch/search/aggregations/bucket/composite/CompositeAggregator.java

```diff
@@ -556,10 +556,13 @@ private void processLeafFromQuery(LeafReaderContext ctx, Sort indexSortPrefix) t
     }
 
     @Override
-    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
-        boolean optimized = filterRewriteOptimizationContext.tryOptimize(ctx, this::incrementBucketDocCount, segmentMatchAll(context, ctx));
-        if (optimized) throw new CollectionTerminatedException();
+    protected boolean tryPrecomputeAggregationForLeaf(LeafReaderContext ctx) throws IOException {
+        finishLeaf(); // May need to wrap up previous leaf if it could not be precomputed
```

Comment:
`finishLeaf` was not getting invoked earlier in case of successful precomputation. Again wondering if that was a bug!?

Reply:
Yeah -- I commented on it here: https://github.com/opensearch-project/OpenSearch/pull/16733#discussion_r1929388793

IMO, this is a potential bug. (Calling `finishLeaf()` redundantly is a no-op.)

Reply:
Got it, just saw that. I am wondering if there is a small test that can reproduce the issue? Although, nothing to do with this PR, just worried that we regressed something and did not even know about that

Reply:
So, I did a bit more digging into it. 

It looks like this wasn't actually a bug, because the two situations were mutually exclusive:

1. `finishLeaf` is a no-op whenever there are no subaggregators (because `finishLeaf` only does something if `leafCtx` is non-null, which only happens when `fillDocIdSet` is true, which only happens when `deferredCollectors` is not `NO_OP_COLLECTOR`, which only happens when there are subaggregators).
2. `canOptimize` is false if there are subaggregators.

So, we were never short-circuiting (and skipping the call to `finishLeaf`) when it mattered.